actionpack/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb

```diff
@@ -10,7 +10,30 @@
 <div id="container">
   <h2><%= h @exception.message %></h2>
 
-  <%= render template: "rescues/_source" %>
-  <%= render template: "rescues/_trace" %>
+  <%= render "rescues/source", source_extracts: @source_extracts, show_source_idx: @show_source_idx, error_index: 0 %>
+  <%= render "rescues/trace", traces: @traces, trace_to_show: @trace_to_show, error_index: 0 %>
+
+  <% if @exception.cause %>
+    <h2>Exception Causes</h2>
+  <% end %>
+
+  <%
+    index = 1
+    while cause = (cause || @exception).cause
+  %>
+    <div class="details">
+      <a class="summary" href="#" style="color: #F0F0F0; text-decoration: none; background: #C52F24; border-bottom: none;" onclick="return toggle(<%= cause.object_id %>)">
+        <%= cause.class.name %>: <%= h cause.message %>
+      </a>
+    </div>
+
+    <div id="<%= cause.object_id %>" style="display: none;">
+      <% wrapper = ActionDispatch::ExceptionWrapper.new(@request.get_header("action_dispatch.backtrace_cleaner"), cause) %>
+      <%= render "rescues/source", source_extracts: wrapper.source_extracts, show_source_idx: wrapper.source_to_show_id, error_index: index %>
+      <%= render "rescues/trace", traces: wrapper.traces, trace_to_show: wrapper.trace_to_show, error_index: index %>
```

Comment:
Why don't we just pass in a wrapper here and in this template? I'm not liking the wrapper instantiation in here.

Perhaps then we could do

```ruby
@wrapper.wrapped_causes.each do |cause_wrapper|
  # …
end
```

Reply:
I thought about this too. I'll do that if we can get agreement on this from more folks.